plugins/omazureeventhubs/omazureeventhubs.c

```diff
@@ -1459,27 +654,119 @@ CODESTARTtryResume
 	DBGPRINTF("omazureeventhubs: tryResume for pData %p returned %d\n", pData, iRet);
 ENDtryResume
 
+BEGINbeginTransaction
+CODESTARTbeginTransaction
+	/* we have nothing to do to begin a transaction */
+ENDbeginTransaction
+
 /* 
- *
+ *	New Transaction action interface
  */
+BEGINcommitTransaction
+	instanceData *__restrict__ const pData = pWrkrData->pData;
+	unsigned i;
+	sbool bDone = 0;
+	protonmsg_entry* pMsgEntry = NULL;
+CODESTARTcommitTransaction
+	pthread_mutex_lock(&pData->mut_doAction);
+#ifndef NDEBUG
+	DBGPRINTF("omazureeventhubs: commitTransaction for pData %p ENTER\n", pData);
+#endif
+	if (pData->bIsConnected == 0) {
+		CHKiRet(setupProtonHandle(pData, 0));
+	}
+
+	// Handle/Expand our proton helper array
+	pData->nProtonMsgs = 0;
+	if (nParams > pData->nMaxProtonMsgs) {
+		// Free old array first!
+		if(pData->aProtonMsgs != NULL) {
+			for(i = 0 ; i < pData->nProtonMsgs ; ++i) {
+				if (pData->aProtonMsgs[i] != NULL) {
+					protonmsg_entry_destruct(pData->aProtonMsgs[i]);
+				}
+			}
+			free(pData->aProtonMsgs);
+		}
+
+		// Expand our proton helper array
+		DBGPRINTF("omazureeventhubs: commitTransaction expand helper array from %d to %d for pData %p\n",
+			pData->nMaxProtonMsgs, nParams, pData);
+		pData->nMaxProtonMsgs = nParams; // Set new MAX
+		CHKmalloc(pData->aProtonMsgs = calloc(pData->nMaxProtonMsgs, sizeof(struct s_protonmsg_entry)));
+	}
+
+	do {
+		// Put unsubmitted messages into Proton
+		for(i = 0 ; i < nParams ; ++i) {
+			// Get reference to Proton Array Helper
+			pMsgEntry = ((protonmsg_entry*)pData->aProtonMsgs[i]);
+			if (	pMsgEntry == NULL) {
+				// Send Message to Proton 
+				writeProton(pData, pParams, i);
+			} else if (	pMsgEntry->submitted == 1 &&
+					pMsgEntry->rejected == 1) {
+				// Reset Message Entry, it will be retried
+				pMsgEntry->submitted = 0;
+				pMsgEntry->accepted = 0;
```

Comment:
Submitted can be accepted or rejected and the different states matter. Could use a bit mask but in this case it is better human readable this way.

Reply:
I thought the live cycle would be
```
unsubmitted -> submitted ----> rejected ---> processed
      ^              |            |
      |              |            |
      +--------------+------------+
```
Is there any situation where the same message is both submitted and rejected? If so, doesn't that mean it would need to be resubmitted, which means change state to unsubmitted and the to submitted again?

Reply:
follow up: in a sense, can't we transit from "rejected" directly to "unsubmitted"?

I assume that on reject a short wait is done for rate-liminting. No matter if this happens during the initial submit or after submit when the storage engine rejects the message.